### ISSUE TITLE
[Discuss] Operating System compatibility of tools and actions

### ISSUE THREAD

**Issue by janpio** [fastlane/fastlane#11859]
The next step in cross-platform (Windows and Linux) support is to find a way to handle code and actions that can not work on a specific platform in a better way than crashing 💣 

For the test suite, we for now solved this by disabling the tests with tags. The tests are not run ("pending test") and the test suite as a whole can pass. But this of course only works for tests - there is no similar mechanism that can warn a user when they use an action in their Fastfile - yet.

--- 

My current idea is to add a list of "operating system compatibility" to each action: 

- known good,
- known bad
- or unknown. 

By default (and if this "compatibility mapping" is missing) all actions are assumed to be "known good" for macOS (they were developed on and for macOS after all) and "unknown" for all other operating systems. 

After manual testing and confirmation, this can manually (Pull Request to the action source code) be changed to "known good" or "known bad" in the code.

Before an action is run, a global "before" method that is run for all tools/actions can check this mapping and output a warning in the console or even block the execution of the tool/action with an error message.

---

This concept is quite similar to how the platform support (on which an action can be used in the `Fastfile`) is defined:

```
def self.is_supported?(platform)
  [:ios, :mac].include?(platform)
end
```

This is used to generate the list of supported platforms in the docs: 

https://github.com/fastlane/fastlane/blob/8a2e328892c34ff5d1d450a759a480cb325c5008/fastlane/lib/assets/ActionDetails.md.erb#L22

But also to output an error if the platform is not listed: 

https://github.com/fastlane/fastlane/blob/8a2e328892c34ff5d1d450a759a480cb325c5008/fastlane/lib/fastlane/runner.rb#L294-L304

---

The definition of compatibility could look like this (pseudocode):

```
is_compatible(operating_system)
  compatibility = { 
    "macos": true, 
    "linux": false, 
    "windows": null 
  }
  return compatibility[operating_system]
end
```

This method, then, could be used in similar ways than `is_supported`: generate documentation on operating system compatibility of tools/actions and error/warning generation at runtime.

(I think it might also be possible to switch the "pending tests" from a static tag written in the test file to also use this compatibility mapping on runtime to decide if a test should be run on the platform or not.)

---

What do you think?

--- 

PS: An additional useful feature could be to be able to define a "message" regarding a "false" value: 

> This action uses Xcode which is not available for this operating system.

The pending tests already do something like that, so it would be neat to keep that. The `compatibility` definition would have to be different then, of course.

**Comment by KrauseFx** [comment]
Sounds great to me, nice write-up 👍

**Comment by markhu** [comment]
As someone who is new to Fastlane, can I take this as a clue that Fastlane is targeted mostly towards developers working locally in MacOS?  In other words, not much of Fastlane will work on a cloud (Linux) Jenkins?

**Comment by janpio** [comment]
Yes and no. Yes, Fastlane is/was focused on macOS. No, not locally but also on cloud servers (e.g. CircleCI). But you should get pretty far with Linux as well (at least as long as you don't require any macOS specific tooling in your actions like Xcode etc.) We are currently working on improving and most importantly documenting that support: https://github.com/fastlane/fastlane/issues/11687

**Comment by janpio** [comment]
After not actively thinking about this for half a year but testing fastlane on Windows and Ubuntu, I think it might make more sense to simplify this to only add a "setting" to the actions that definitely do not work on a platform. No default_compatibility and just an optional is_incompatible list of platforms where an action definitely does not work (yet).

First implementation of this in https://github.com/fastlane/fastlane/pull/13703

**Comment by janpio** [comment]
Ok, https://github.com/fastlane/fastlane/pull/13703 - `is_incompatible` is now somehow fleshed out and actually working.

**Comment by Marla3** [comment]
An issue to keep in mind is compatibility between different versions of the same item of software. Software is regularly upgraded, many on average every 2-3 years. Between major upgrades, minor upgrades or 'service packs' can be released, mainly intended to correct problems or bugs. It is important to manage this process, known as 'version control', to ensure that all users are equipped with the same version of the software. If different users have different versions, compatibility problems can arise, even within the same software.
